File: algorithms/algorithms-cplusplus/leetcode/Question_0530_Minimum_Absolute_Difference_in_BST.cc
#include <limits.h>
#include <algorithm>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  int getMinimumDifference(TreeNode* root) {
    int res = INT_MAX, pre = -1;
    help(root, pre, res);
    return res;
  }

  void help(TreeNode* root, int& pre, int& res) {
    if (!root) {
      return;
    }
    help(root->left, pre, res);
    if (pre != -1) {
      res = min(res, root->val - pre);
    }
    pre = root->val;
    help(root->right, pre, res);
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0530_Minimum_Absolute_Difference_in_BST.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0530_Minimum_Absolute_Difference_in_BST {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private int res;
    private Integer pre;

    public int getMinimumDifference(TreeNode root) {
        if (root == null) {
            return -1;
        }
        res = Integer.MAX_VALUE;
        Integer pre = null;
        help(root);
        return res;
    }

    private void help(TreeNode root) {
        if (root == null) {
            return;
        }
        help(root.left);
        if (pre != null) {
            res = Math.min(res, root.val - pre);
        }
        pre = root.val;
        help(root.right);
    }
}
